include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -397,33 +397,52 @@ class __buffer_impl
     }
 };
 
-template <typename _T>
 struct __sycl_usm_free
 {
     sycl::queue __q;
 
     void
-    operator()(_T* __memory) const
+    operator()(void* __memory) const
     {
-        sycl::free(__memory, __q.get_context());
+        sycl::free(__memory, __q);
     }
 };
 
 template <typename _T, sycl::usm::alloc __alloc_t>
-struct __sycl_usm_alloc
+_T*
+__sycl_usm_alloc(const sycl::queue& __q, std::size_t __elements)
 {
-    sycl::queue __q;
+    if (_T* __buf = sycl::malloc<_T>(__elements, __q, __alloc_t))
+        return __buf;
 
-    _T*
-    operator()(::std::size_t __elements) const
-    {
-        if (auto __buf =
-                static_cast<_T*>(sycl::malloc(sizeof(_T) * __elements, __q.get_device(), __q.get_context(), __alloc_t)))
-            return __buf;
+    throw std::bad_alloc();
+}
```

Comment:
I don't love that `__sycl_usm_alloc` and `__allocate_usm` both exist, and seem to serve very similar purposes with similar interfaces.  I understand the idea is to keep `__sycl_usm_alloc` temporarily to support `result_and_scratch_storage` until it goes away.  However, the changes made force us to touch all usages anyway.  
Can we instead just replace those usages with the new `__allocate_usm`?

Alternatively, if we dont want to repeat `nullptr` checks in the calling code, another option could be to make `__sycl_usm_alloc` wrap `__allocate_usm` and throw if `nullptr` return.

At the very least I'd want some comment saying this is only here for legacy support and will be going away.